bindgen-tests/tests/expectations/tests/macro_const.rs

```diff
@@ -5,10 +5,16 @@
     non_upper_case_globals
 )]
 
-pub const foo: &[u8; 4usize] = b"bar\0";
-pub const CHAR: u8 = 98u8;
-pub const CHARR: u8 = 0u8;
-pub const FLOAT: f64 = 5.09;
-pub const FLOAT_EXPR: f64 = 0.005;
+pub const foo: *const ::std::os::raw::c_char = {
+    const BYTES: [u8; 4] = *b"bar\0";
+    BYTES.as_ptr() as *const ::std::os::raw::c_char
+};
+pub const CHAR: ::std::os::raw::c_char = 'b' as ::std::os::raw::c_char;
+pub const CHARR: ::std::os::raw::c_char = '\0' as ::std::os::raw::c_char;
+pub const FLOAT: f32 = 5.09;
+pub const FLOAT_EXPR: f32 = 0.005;
 pub const LONG: u32 = 3;
-pub const INVALID_UTF8: [u8; 5usize] = [240u8, 40u8, 140u8, 40u8, 0u8];
+pub const INVALID_UTF8: *const ::std::os::raw::c_char = {
```

Comment:
In C, a string literal is a `const *char`, so in Rust it should be the same.

I guess using `&CStr` will make more sense once all needed functions are `const`-stable.

Reply:
Hmm I checking the standard and it says
> In translation phase 7, a byte or code of value zero is appended to each multibyte character sequence that results from a string literal or literals. The multibyte character sequence is then used to initialize an array of static storage duration and length just sufficient to contain the sequence.

I understand that for any evaluation such as `#if STR_LIT_1 == 0` , the literal would get converted into a pointer but from the standard I would understand that they are treated as arrays first. What do you think?

Reply:
> they are treated as arrays first

But what's the difference? Aren't arrays just syntactic sugar for pointer arithmetic?

> What do you think?

I guess in the end using `&CStr` would probably be the most ergonomic, since it provides an `as_ptr` as well as `to_str` method. Currently this would require `unsafe` because only `[from_bytes_with_nul_unchecked](https://doc.rust-lang.org/core/ffi/struct.CStr.html#method.from_bytes_with_nul_unchecked)` is `const`-stable so far.

Reply:
Well arrays (or even just slices) intrinsically have more information because they can provide their size which is something that rust users could be relying on and we would be removing that part from the API by using `*const c_char`

Reply:
I think my only concern is `bindgen` currently allows denying unsafe code, but I guess we can just use `&CStr` and annotate all strings with `#[allow(unsafe_code)]`.

Reply:
Okay, what should happen in this edge case?

```c
#define HELLO1 "WORLD"
#define HELLO2 (char* const)"WORLD"
```

Should `HELLO2` be a pointer because of the explicit cast or also a `&CStr`?

Reply:
I have made both a `&CStr` now. That way you can call `as_ptr()` or `to_bytes_with_nul()` depending on your use case.